llvm/include/llvm/Transforms/Vectorize/SandboxVectorizer/DependencyGraph.h

```diff
@@ -29,22 +29,42 @@
 
 namespace llvm::sandboxir {
 
+class DependencyGraph;
+class MemDGNode;
+
+/// SubclassIDs for isa/dyn_cast etc.
+enum class DGNodeID {
+  DGNode,
+  MemDGNode,
+};
+
 /// A DependencyGraph Node that points to an Instruction and contains memory
 /// dependency edges.
 class DGNode {
+protected:
   Instruction *I;
+  // TODO: Use a PointerIntPair for SubclassID and I.
+  /// For isa/dyn_cast etc.
+  DGNodeID SubclassID;
   /// Memory predecessors.
   DenseSet<DGNode *> MemPreds;
-  /// This is true if this may read/write memory, or if it has some ordering
-  /// constraints, like with stacksave/stackrestore and alloca/inalloca.
-  bool IsMem;
+
+  DGNode(Instruction *I, DGNodeID ID) : I(I), SubclassID(ID) {}
+  friend class MemDGNode; // For constructor.
 
 public:
-  DGNode(Instruction *I) : I(I) {
-    IsMem = I->isMemDepCandidate() ||
-            (isa<AllocaInst>(I) && cast<AllocaInst>(I)->isUsedWithInAlloca()) ||
-            I->isStackSaveOrRestoreIntrinsic();
+  DGNode(Instruction *I) : I(I), SubclassID(DGNodeID::DGNode) {
+    assert(!isMemDepCandidate(I) && "Expected Non-Mem instruction, ");
   }
+  DGNode(const DGNode &Other) = delete;
+  virtual ~DGNode() = default;
+  /// \Returns true if \p I is a memory dependency candidate instruction.
+  static bool isMemDepCandidate(Instruction *I) {
+    return I->isMemDepCandidate() ||
+           (isa<AllocaInst>(I) && cast<AllocaInst>(I)->isUsedWithInAlloca()) ||
```

Comment:
dyn_cast!

Reply:
The alternative is the following which is not as readable as the original code.
```
if (I->isMemDepCandidate())
  return true;
auto *AllocaI = dyn_cast<AllocaInst>(I);
return AllocaI != nullptr ? AllocaI->isUsedWithInAlloca() || I->isStackSaveOrRestoreIntrinsic();
```

Reply:
```
if (I->isMemDepCandidate())
  return true;
if (auto *AllocaI = dyn_cast<AllocaInst>(I))
  if (AllocaI->isUsedWithInAlloca())
    return true;
return I->isStackSaveOrRestoreIntrinsic();
```

Reply:
Yeah, the original one is much better.

Reply:
But it is a violation of the coding style guide.

Reply:
It might be, but isn't more readable code more important?

Reply:
Maybe we should start a discussion thread about this pattern because it does make this particular code pattern easier to read, which is probably why it's widely used quite a lot throughout the codebase. But on the other hand we should not have code that does not follow the coding standard. So either the coding standard should make it clear that this is acceptable, or specifically discourage it. Wdyt ?

Reply:
"Note that you should not use an isa<> test followed by a cast<>, for that use the dyn_cast<> operator."
https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates

Reply:
We can argue about the `should` but ....

Reply:
Let me start a thread on discourse about it. It would be good to know what the rest of the community thinks about it.

Reply:
https://discourse.llvm.org/t/coding-style-isa-cast-pattern/81473

Reply:
Let me change it to what @david-xl suggested in the thread, which is still readable but not too verbose:
```
AllocaInst *Alloca;
return I->isMemDepCandidate() || 
      ((Alloca = dyn_cast<AllocaInst>(I)) && Alloca->isUsedWithInAlloca()) ||
      I->isStackSaveOrRestoreIntrinsic();
```